inst/tests/tests.Rraw

```diff
@@ -16846,3 +16846,15 @@ A = data.table(A=c(complex(real = 1:3, imaginary=c(0, -1, 1)), NaN))
 test(2138.3, rbind(A,B), data.table(A=c(as.character(A$A), B$A)))
 A = data.table(A=as.complex(rep(NA, 5)))
 test(2138.4, rbind(A,B), data.table(A=c(as.character(A$A), B$A)))
+
+# Nested calls to data.table() do not avoid shallow copies #1629
+add_col1 = function(dt) {
+  if (is.data.table(dt)) dt[, new_col:=NA]
+  if (is.list(dt)) lapply(dt, add_col1)
+  invisible()
+}
+dt = data.table(a=c(1,2), b=list(data.table(d=c("a", "b"), e=c(100, 200))))
+add_col1(dt)
```

Comment:
Using v1.12.2 (see https://github.com/Rdatatable/data.table/issues/1629#issuecomment-846121806) I ran the new test to confirm it fails correctly.  New tests .2 and .3 fail (good), .1 passes (also fine as we often have extra tests in a set of tests which test the test setup).  However, the `add_col1(dt)` produces warnings in v1.12.2 which aren't being checked are not produced.  So it would go the extra mile to wrap the `add_col1()` call with `test()` so that it is checked that no warnings are produced by it (it doesn't produce warnings in v1.12.4 onwards). 
```
> add_col1(dt)
Warning messages:
1: In `[.data.table`(dt, , `:=`(new_col, NA)) :
  Invalid .internal.selfref detected and fixed by taking a (shallow) copy of the data.table so that := can add this new column by reference. At an earlier point, this data.table has been copied by R (or was created manually using structure() or similar). Avoid names<- and attr<- which in R currently (and oddly) may copy the whole data.table. Use set* syntax instead to avoid copying: ?set, ?setnames and ?setattr. If this message doesn't help, please report your use case to the data.table issue tracker so the root cause can be fixed or this message improved.
2: In `[.data.table`(dt, , `:=`(new_col, NA)) :
  Invalid .internal.selfref detected and fixed by taking a (shallow) copy of the data.table so that := can add this new column by reference. At an earlier point, this data.table has been copied by R (or was created manually using structure() or similar). Avoid names<- and attr<- which in R currently (and oddly) may copy the whole data.table. Use set* syntax instead to avoid copying: ?set, ?setnames and ?setattr. If this message doesn't help, please report your use case to the data.table issue tracker so the root cause can be fixed or this message improved.
```

Also the test could be moved into 2049 to be next to the other one.

I'll do this and then merge.